### ISSUE TITLE
interchangable correlation functions

### ISSUE THREAD

**Issue by d-chambers** [eqcorrscan/EQcorrscan#119]
I want to discuss a few changes I am working on before I get too far along with the implementation because it has the potential to break the current API of a few function.

The purpose is to allow EQcorrscan users to switch between different correlation implementations without having to manually change the codebase.  This would allow much greater flexibility in how EQcorrscan can be used. For example, users running on a distributed system using SLURM may want to use the fftw/openMP stuff while those running on a single machine may find it more efficient to simply use threading with numpy. Moreover, it may be of interest to some users to plug in their own implementations. For example,  I have been think of writing a GPU correlation function using Numba, but it certainly wont be the "right" implementation in all cases, even if it runs much faster, because it may require fairly specific hardware (eg NVIDIA GPUs).  

In order to do this we can use the fact that most of the functions in  EQcorrscan.utils.correlate have compatible signatures of the form:

```python
def func(templates: np.ndarray, stream: np.ndarray, pads: np.ndarray) -> np.ndarray:
    pass
```

So we can define a single gatekeeper function, a [facade](http://python-3-patterns-idioms-test.readthedocs.io/en/latest/ChangeInterface.html#facade) of sorts, which all other functions/classes in EQcorrscan will call to compute correlations. The function might have the following form:

```python
def normxcorr(templates, stream, pads, *args, **kwargs):
    # some custom logic here to get the correct correlation function
    func = get_corr_func(name=kwargs.get('correlation_function')) 
    return func(templates, stream, pads, *args, **kwargs)
```
This would allow the user to change between the different functions when calling `normxcorr`, but would provide a reasonable default optimized for the general case. 

Moreover, we could allow users to swap out the general correlation function (backend) for a different correlation function included in EQcorrscan or one that the user defines. This could be done by permanently modifying the global context, or using a context manager similar to [dask's global context module](https://github.com/dask/dask/blob/master/dask/context.py). It might look like this:

```python
from eqcorrscan import normxcorr, set_correlation_func
# get templates, stream, pads
...
# use default correlation back-end
cc = normxcorr(templates, stream, pads)

# temporarily swap the correlation backend for cool_new_normxcorr
with set_correlation_func(cool_new_normxcorr):
    cc = normxcorr(templates, stream, pads)
# the backend gets changed back when the context manager exits 

# permanently (for this session) swap out the correlation backend for cool_new_normxcorr
set_correlation_func(cool_new_normxcorr)
cc = normxcorr(templates, stream, pads)
```
We can then use args and kwargs to allow function specific parameters, when necessary. 

If we take this one step further, and this is the part that could break the existing API, we could register a bunch of functions, some created using [partials](http://chriskiehl.com/article/Cleaner-coding-through-partially-applied-functions/) or other higher order functions, to define common variations of similar functions. 

For example, we might register a bunch of functions like this:

```python
from functools import partial

from eqcorrscan import register_xcorr
from eqcorrscan.utils.correlate import numpy_normxcorr, time_multi_normxcorr, fftw_multi_normxcorr

def multi_thread(func):
    # make a function multi-threaded

def multi_processes(func):
    # make a function multiprocess

# register functions 
register_xcorr(numpy_normxcorr, 'numpy')
register _xcorr(multi_processes(numpy_normxcorr), 'numpy_multiprocess')
register _xcorr(multi_thread(numpy_normxcorr), 'numpy_multithread')
register_xcorr(time_multi_normscorr, 'time_domain')
register_xcorr(partial(fftw_multi_normxcorr, threaded=False), 'fftw')
register_xcorr(partial(fftw_multi_normxcorr, threaded=True), 'fftw_threaded')
# etc
```
Then longer functions that spend many lines of code trying to figure out which correlation function to use could accept a single parameter defining the name of the correlation function (`None` would use the default). The signature of `multichannel_normxcorr`, for example, could be changed from: 

```python
def multichannel_normxcorr(templates, stream, cores=1, time_domain=False, 
                                              openmp=False):
    pass
```
to this:

```python
def multichannel_normxcorr(templates, stream, correlation_function=None):
    pass
```
Then a single line of code could replace about 50 lines that include a bunch of if else statements to figure out which correlation function to use.

```python
def multichannel_normxcorr(templates, stream, correlation_function=None):
    # prep streams, templates, pads
    ...
    return normxcorr(tempaltes, stream, pads, correlation_function=correlation_function) 
```

I have most of the implementation worked out in my head, and a good start [here](https://github.com/d-chambers/EQcorrscan/tree/correlate_refactor) but I wanted to get some feedback before I proceed much further.

**Comment by calum-chamberlain** [comment]
That all sounds really good - I will have to think a bit about it and get back to you. @cjhopp should have a think too.

I was thinking of something similar, but didn't have the understanding to make it happen, so in principle keen.  I was also imagining that a user could run some kind of setup-test to work out which function would be best given their architecture and problem scale, but I don't know how we could do that. I imagined something run by setup.py to generate a look-up table with optimum functions for given problem sizes, but it might end up too clumsy and abstracted.

**Comment by d-chambers** [comment]
In theory it would be something like fftw "wisdom" right? The implementation would get a bit complicated, and just another thing that could go wrong when a new user is trying to install the package. I vote on not doing it in order to keep the installation as simple as possible.

That being said, a function lets the user provide a "typical stream", runs all known correlations functions on it, then returns a dictionary of how long each function took (maybe return None if it fails) could be useful in guiding the user that is actively thinking about (and actually in need) of optimization.

**Comment by calum-chamberlain** [comment]
Yeah, something like that, but yes, agree on keeping install simpler.

The function stream may be something I think about alongside the openMP work.

As for the main ideas in the issue:
I had a look through your branch - it looks good - I like the wrapper and I can see that it will clean up the code a lot and be really good for usability. I would go for it. Regarding the partials - I think I understand where you are going with that, and I think it is a good way to go to get around the threaded/non-threaded fftw.  I am not overly concerned with changing the API of these relatively internal (and new) functions, if it was say, an extra one or two lines of code to just raise some kind of warning saying your are ignoring old arguments if they are given as kwargs, then it might be worth putting those in for a release cycle (but with it being to internal functions that people are unlikely to have played with, I wouldn't be concerned if it wasn't there).

Can you see any other issues that you might run in to? I think the idea is great, and should be done - it would be really nice to have a simple means of changing the backend (and I like the idea of setting it, to me it feels like setting the plotting backend of matplotlib, but probably not similar in practice), I also think that if we can allow people a simple way to extend with their 'new_cool_function' that would be awesome.  I would love to see a GPU implementation in there.

I should be reminded that I need to fix up the time domain cross-correlation if this goes ahead, it is not properly normalised as it stands, which is why the tests have more flexibility when comparing to time-domain results - it works well for short correlations (e.g. one sample), but isn't correct for longer time-series with variable moving-window means. It should be simple to do, I just need to do it, but hadn't thought I needed to because it was quite buried.

**Comment by calum-chamberlain** [comment]
Also, the default for cross_chan_coherance should probably stay as time-domain, for those small correlations it is usually faster - and the parallel loop call might be worth keeping in there.

**Comment by d-chambers** [comment]
Ok,

I will go ahead and see what I can come up with. 

As far as technical challenges I only see one so far: The interface for `fftw_multi_normxcorr` is different than the other functions because it is pushing some looping down to the C level. I will need to think about how to handle the difference. Maybe I should wrap the other functions in python loops to match this signature? 

For `cross_chan_coherence` I will probably end up setting a parameter that will use the multiprocessing version of the time domain function by default, but let the user change that if desired. I could imagine one might be interested in correlating many large chunks against each other looking for repeating earthquakes without known templates, which may make the ability to switch the xcorr function advantageous. 

Let me think about it a bit more and prototype some things. I will ping you in this thread if I have questions or need advice.

**Comment by calum-chamberlain** [comment]
That sounds excellent.  

You make a good point about the difference with `fftw_multi_normxcorr`. If you were to wrap the others in Python loops, would you have to implement a multiprocesing and a serial loop, or would you be able to handle that with partials?

Thanks!

**Comment by d-chambers** [comment]
I have two ideas:

1. 

We do something similar to (singledispatch)[https://www.python.org/dev/peps/pep-0443/] which is in python 3's standard library (functools). This will allow us to register multiple interfaces with a function or simply use defaults. I am thinking about the implementation but this is what I would like it to look like:

```python
from eqcorrscan import register_normxcorr

@register_normxcorr
def normxcor1(templates, stream, pads):
    ...

@register_normxcorr
def normxcor2(templates, stream, pads):
    ...

@normxcor2.register_multiprocess
def _normxcorr2_multiprocess(template_array, stream_array, pad_array, seed_ids):
   ...
```

The decorator will automatically give the function standard multiprocess/threading interfaces (as with normcorr1) but we could manually register a non-standard (as was done with normcorr2)

I am not sure if they should be triggered with an attr like access or using kwargs

```python 
# this would access normcorr2
cc = normcorr2(templates, stream, pads)

# this could work for dispatching to _normxcorr2_multiprocess
cc_multiprocess = normxcorr1.multiprocess(template_array, stream_array, pad_array, seed_ids)

# this could also work for dispatching to _normxcorr2_multiprocess
cc_multiprocess = normxcorr1(template_array, stream_array, pad_array, seed_ids, method='multiprocess')
```

 2.

We could just keep the interfaces separate. One function for accessing the (templates, stream, pads) interface and another for the (template_array, stream_array, pad_array, seed_ids) interface. We would also need separate register functions. 

```python
def normxcorr(templates, stream, pads, *args, **kwargs):
    func = get_normxcorr_func()
    return func(templates, stream, pads, *args, **kwargs)


def multichannel_normxcorr(template_array, stream_array, pad_array, seed_ids, *args, **kwargs):
    func = get_multichannel_normxcorr_func()
    return func(template_array, stream_array, pad_array, seed_ids, *args, **kwargs)
``` 

Which do you prefer?

**Comment by calum-chamberlain** [comment]
I think I prefer the first option, with the attribute, rather than the kwargs (it just seems more obvious that it is different and therefore requires different arguments).
I was leaning towards the second one, but, if the first option can be made to work with py2.7 (definitely starting to wish I hadn't set up to support it), and provides that clean of an interface, and isn't too much work, then I would go with that I think.
Which do you prefer?

**Comment by cjhopp** [comment]
I'll get out of the way, but I'm excited about the flexibility this is going to create.

**Comment by d-chambers** [comment]
Ok I think I have this working. 

There are two interfaces to keep in mind as you read this: array_xcorr and stream_xcorr.

```python
from typing import List, Tuple

import numpy as np
from obspy import Stream

# array_xcorr interface, eg numpy_normxcorr 
def array_xcorr(templates_array: np.ndarray , stream_array: np.ndarray, pad_array: np.ndarray, *args, 
                             **kwargs) -> (np.ndarray, np.ndarray):

# stream_xcorr interface eg multichannel_normxcorr
def stream_xcorr(templates: List[Stream], stream: Stream) -> (List[np.ndarray], List[int], List[Tuple[str, str]]):

```
Here is the summary of what I have done so far:

1. xcorr functions can now be registered with the `eqcorrscan.utils.correlation.register_array_xocrr` function. This function can act as a decorator with no arguments, a decorator with arguments, or a callable. For example:

```python
from eqcorrscan.utils.correlate.register_array_xcorr

# option 1
@register_array_xcorr
def new_corr_func1(template_array, stream_array, pads):
    pass

# option 2
@register_array_xcorr('func2')
def new_corr_func2(template_array, stream_array, pads):
    pass

# option 3
def new_corr_func3(templates_array, stream_array, pads):
    pass

register_array_xcorr('func3', new_corr_func3)
```

2. registered xocrr functions have standard concurrency wrappers added as function attributes.
This is intended to abstract the use of pools (both process and threads) so less code is repeated. The design idea is loosely based on single dispatch of the functools module in python 3. 
Each registered function now has "multithread", "multiprocess", and "concurrent" methods (or functions? not sure what to call functions attached to functions). These functions must implement the stream_xcorr interface. 

```python
from eqcorrscan.utils.correlate import numpy_normxcorr

# function to run numpy with multiple threads (this works ok because numpy releases the GIL)
numpy_normxcorr.multithread

# function to run numpy xcorr with multiple processes
numpy_normxcorr.multiprocess

# concurrent is just threaded by default, but can be customized
```

Alternative concurrency methods can be registered for each function the same way single dispatch registers branches of generic functions. 

```python
from eqcorrscan.utils.correlate import register_array_xcorr

@register_array_xcorr
def array_xcorr(template_array, stream_array, pad_array):
    ...

# register a new way of implementing multiprocessing
@array_xcorr.register('multiprocessing')
def _new_multiprocess(template_streams, stream, *args, **kwargs):
    ...

# now calling array_xcorr.multiprocess will call _new_multiprocess
```

3. `get_array_xcorr` and `get_stream_xcorr` abstract all of this away to make an easy interface for other parts of eqcorrscan to get the appropriate function based on an existing xcorr function tag or a new callable

```python
from eqcorrscan.utils.correlate import get_array_xcorr, get_stream_xcorr

# get a multi-threaded version of numpy function
func = get_stream_xcorr('numpy', concurrency='multithread')

# get a single threaded version of time_domain 
func = get_stream_xcorr('time_domain')

# get fftw (all concurrency is overwritten to be based on fftw_multi_normxcorr)
func = get_stream('fftw', concurrency='concurrent')

# fftw is the default so this also returns fftw based on fftw_multi_normxcorr
func = get_stream_xcorr()

def new_func(template_array, stream_array, pad_array):
    pass

# this returns a stream_xcorr function that uses new_func in a thread pool, 
# and registers new_func as an array_xcorr function
func = get_stream_xcorr(new_func, 'multithread')
````

However, the default stream_xcorr function can be permanently (at least for a python session) or temporarily changed with the set_xcorr instance of _Context.
```python
from eqcorrscan.utils.correlate import get_array_xcorr, set_xcorr

func = get_array_xcorr()  # returns fftw, the "default default"

# permanently change default xcorr to time_domain
set_xcorr('time_domain') 
func = get_array_xcorr()   # returns single-threaded stream_xcorr that uses time_multi_normxcorr

# temporarily use numpy as default
with set_xcorr('numpy'):
    func = get_array_xcorr()   # returns single-threaded stream_xcorr that uses numpy

func =  get_array_xcorr()  # again returns single threaded time_domain version
```

As promised, this allows the user to swap out the xcorr back end for all parts of eqcorrscan  that use  `get_array_xcorr` that do not explicitly ask for a particular xcorr function.

4. I added replaced `multichannel_normxcorr` with a function returned by `get_stream_xcorr`, and added appropriate arguments to match_filter. 

5. I was having a hard time expressing the tests I wanted to write in the unittest style so I rewrote eqcorrscan/test/correlate_test.py using pytest idioms. 

6. I added a few session fixtures in conftest.py to clean up the doctest droppings (temp files/directories).

7. Pytest seems to have borked some of the pep8 stuff so there are other misc. changes I had to make in order to get the pep8 tests to pass (still working on these actually)

Take a look at the correlate_refactor branch on github.com/d-chambers/EQcorrscan and let me know what you think. I believe I branched off master so I may need to rebase with develop? I have some pointless commits too that I will try to clean up and combine with others. I am fairly new to the more complicated git stuff, as well as all the CI, sphinx doc generation, etc.  so any tips on moving forward are appreciated.

**Comment by calum-chamberlain** [comment]
That all sounds pretty excellent.  I have had a read through of your utils.correlate and it looks really nice - I especially like the `_pool_boy` (both for the name, and the simplicity of it). Good work.

I haven't had a look at other files, is there anything else in there that needs a look?

If you can rebase from develop that would be ideal, but it's not the end of the world if not, we can do some backwards merging if we need to (but it would be cleaner to come off develop).  The way things are progressing, I'm going to release a small patch release (0.2.5, mostly to cope with an annoying bug in lag-calc #128 where I wasn't checking the stream length properly) in the next day or so, so master will get updated then.  It might be worth thinking about having this for a 0.2.6 release (or a 0.3.0, it is much more than a patch).

Regarding pep8 things - I have recently added stickler ci for pep8 checks in develop, and removed those checks from pytest - if you rebase on develop those changes will come through to you.

I am happy for you to use pytest idioms, unittest doesn't fit well with a lot of things, but it is what I know - hopefully through your tests I will learn more pytest styling.

Regarding the sphinx docs, you can make the docs locally by moving to the eqcorrscan/doc directory and running `make html` (bash) - this will generate the webpages in the eqcorrscan/doc/build directory and you can check them there using a web-browser, but I wouldn't worry about that until we open a PR, from there we can get readthedocs to build the branch and we can check a few things out.

**Comment by calum-chamberlain** [comment]
Closed by #139
